File: pyramid.py
class Pyramid:
    "Pyramid class for fastest slide problem"
    
    def __init__(self, pyramid_path=None):
        if(pyramid_path == None):
            self.layers = []
            return

        with open(pyramid_path, "r") as f:
            for index, line in enumerate(f):
                if(index == 0):
                    self.layers = [None] * int(line)
                else:
                    self.layers[index-1] = [int(number) for number in line.rstrip().split(" ")]

    def addLayer(self, layer):
        if(len(layer) == 0):
             raise Exception("Layer length must be greater than 0")
        
        if(len(self.layers) == 0):
            self.layers.append(layer)
            return

        if(len(layer) != len(self.layers[-1]) + 1):
             raise Exception("Layer length must be one more than the previous layer")
        
        self.layers.append(layer)

    def removeLayer(self):
        if(len(self.layers) == 0):
            raise Exception("No layers to remove")
        return self.layers.pop()
        
    def getFastestSlide(self):
        layers_copy = self.layers.copy()
        #start for second to layer and go backwards
        for layer_index in range(len(layers_copy)-2, -1, -1):
            current_layer = layers_copy[layer_index]
            last_layer = layers_copy[layer_index+1]
           
            for number_index in range(len(current_layer)):
                current_layer[number_index] += min(last_layer[number_index], last_layer[number_index + 1])

        return layers_copy[0][0]

File: pyramid.js
'use strict';
const fs = require('fs');

module.exports = class Pyramid {
    constructor(filePath = "") {
        this.pyramidLayers = [];
        if (!filePath) {
            return;
        }

        fs.readFileSync(`${__dirname}/${filePath}`, 'utf8').split('\n').forEach((line, index) => {
            if (index === 0 || line.length === 0) {
                return;
            }
            const layer = line.split(' ').map(Number)
            this.addLayer(layer)
        });
    }

    addLayer(layer) {
        if (layer.length === 0) {
            throw new Error("Layer length must be greater than 0")
        }

        if (this.pyramidLayers.length === 0 && layer.length !== 1) {
            throw new Error("First layer must have a length of 1")
        }

        if (this.pyramidLayers.length !== 0 && layer.length !== this.pyramidLayers[this.pyramidLayers.length - 1].length + 1)
            throw new Error("Layer length must be one more than the previous layer")

        this.pyramidLayers.push(layer);
    }

    removeLayer() {
        return this.pyramidLayers.pop()
    }

    getFastestSlide() {
        if (this.pyramidLayers.length === 0) {
            return undefined;
        }

        if (this.pyramidLayers.length === 1) {
            return this.pyramidLayers[0][0];
        }

        //For each layer, take the minimum of every numbers children and add it to itself until at the top of the pyramid
        const reduced = this.pyramidLayers.reduceRight((lastLayer, currentLayer, index) => {

            //Skip the last row of the pyramid
            if (index === this.pyramidLayers.length - 1)
                return lastLayer;

            return currentLayer.map((number, numberIndex) => {
                const leftChild = lastLayer[numberIndex]
                const rightChild = lastLayer[numberIndex + 1];
                return number + Math.min(leftChild, rightChild)
            })
        });

        return reduced[0];
    }
}